rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala

```diff
@@ -198,43 +208,97 @@ class ReplayRSpace[F[_], C, P, E, A, R, K](store: IStore[F, C, P, A, K], branch:
+                                             c -> {
+                                               if (c == channel)
+                                                 Seq((Datum(data, persist, produceRef), -1))
+                                               else as
+                                             }
                                            }
                                          }
-                                       }
-            result <- extractFirstMatch(channels, matchCandidates, channelToIndexedDataList.toMap) match {
-                       case Right(None)             => runMatcher(comm, produceRef, remaining)
-                       case Right(produceCandidate) => produceCandidate.pure[F]
-                       case Left(_)                 => ???
-                     }
-          } yield result
+              result = extractFirstMatch(channels, matchCandidates, channelToIndexedDataList.toMap) match {
+                case Right(None)             => Accumulator(remaining, None)
+                case Right(produceCandidate) => Accumulator(remaining, produceCandidate)
+                case Left(_)                 => ???
+              }
+            } yield result
+        }
+      concurrent
+        .iterateWhileM(Accumulator(groupedChannels.toList))(go)(
+          acc => acc.channels.nonEmpty && acc.result.isEmpty
+        )
+        .map(_.result)
+    }
 
+    @SuppressWarnings(Array("org.wartremover.warts.Throw"))
+    // TODO stop throwing exceptions
+    def getCommOrProduceCandidate(
+        comms: Seq[COMM],
+        produceRef: Produce,
+        groupedChannels: Seq[Seq[C]]
+    ): F[Either[COMM, ProduceCandidate[C, P, R, K]]] = {
+      type COMMOrProduce = Either[COMM, ProduceCandidate[C, P, R, K]]
+      type Accumulator   = (Seq[COMM], Option[COMMOrProduce])
+      implicit class RichAccumulator(acc: Accumulator) {
+        def comms: Seq[COMM]              = acc._1
+        def result: Option[COMMOrProduce] = acc._2
+      }
+      object Accumulator {
+        def apply(comms: Seq[COMM], cp: Option[COMMOrProduce] = None): Accumulator = (comms, cp)
       }
+      def go(acc: Accumulator): F[Accumulator] =
+        acc.comms match {
+          case Nil =>
+            val msg = "comms must not be empty"
+            logger.error(msg)
+            throw new IllegalArgumentException(msg)
+          case commRef :: Nil =>
+            runMatcher(commRef, produceRef, groupedChannels) map {
+              case Some(x) => Accumulator(Nil, Some(Right(x)))
+              case None    => Accumulator(Nil, Some(Left(commRef)))
+            }
+          case commRef :: rem =>
+            runMatcher(commRef, produceRef, groupedChannels) map {
+              case Some(x) => Accumulator(rem, Some(Right(x)))
+              case None    => Accumulator(rem, None)
+            }
+        }
+      concurrent.iterateWhileM(Accumulator(comms))(go)(_.result.isEmpty).map(_.result.get)
```

Comment:
same thing here, tailrecM with `f: Seq[COMM] => Either[Seq[COMM], COMMOrProduce]`

one can think of a type `Iter[A, B] = A => Either[A, B]` which means "a function to be called with accumulator `A` until it returns a `B` and is returning the new `A` otherwise. This is exactly what tailrecM takes as its f argument